JDBC Client with Inbuilt Pool Driver.

PradeepSurale opened the pull request:
## Changes proposed in this pull request

- JDBC Client with inbuilt Connection Pooling feature.
- JDBC Pool Client Driver is identified by the URL parameter and Driver Class
e.g. jdbc:snappydata:pool://localhost:1527/ Or jdbc:snappydata:thrift:pool://localhost:1527/
  Driver Class : "io.snappydata.jdbc.ClientPoolDriver"
- For pooling purpose we have integrated the tomcat connection pool library.

## Patch testing

- Tested with independent spark cluster with spark.read.jdbc(...) APIs
  e.g 
    (1)
    val df = spark.read.jdbc(JDBC_URL, "AIRLINE", getProperties)
    val result = df.select("flightnum").groupBy("flightnum").avg("flightnum")
    result.show()
    (2)
    val pushdown_query = s"(select * from AIRLINE where flightnum=$i limit 2) emp_alias"
    val df = spark.read.jdbc(url=JDBC_URL, table=pushdown_query, properties=getProperties)
    df.show()
    (3)
    val df = spark.read.jdbc(url=JDBC_URL,
      table="AIRLINE",
      columnName="YEAR_",
      lowerBound=0,
      upperBound=100,
      numPartitions=16,
      connectionProperties=getProperties)
    df.show() 
    
- Tested via ScalaTestSuite with Java Base JDBC APIs
- Manual Testing for the Pool creation and validation etc.

A formal test suits needs to be check-in for this.

## Is precheckin with -Pstore clean?
- Needs to be tested.

## ReleaseNotes changes

(Does this change require an entry in ReleaseNotes? If yes, has it been added to it?)

## Other PRs 

(Does this change require changes in other projects- snappydata, spark, spark-jobserver, aqp? Add the links of PR of the other subprojects that are related to this change)

PradeepSurale committed:
JDBC Client Inbuilt Pool Driver.

Added a ClientPoolDriver with connection pooling feature.
Internally we have used the Tomcat connection pooling library.

kneeraj commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/ClientPoolDriver.java:
```diff
@@ -0,0 +1,174 @@
+    public Connection connect(String url, Properties properties) throws SQLException {
+
+        if (!acceptsURL(url)) {
+            System.out.println("Returning null connection From ClientPoolDriver ");
```
Remove this println.

kneeraj commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/ClientPoolDriver.java:
```diff
@@ -0,0 +1,174 @@
+        }
+
+        properties = (properties == null) ? getPoolProperties(url) : properties;
+        properties.setProperty(TomcatConnectionPool.POOL_URL, url);
```
I think we should use the parent driver for pool. Not now but think on it and change before merge to master.

kneeraj commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/ClientPoolDriver.java:
```diff
@@ -0,0 +1,174 @@
+        } else {
+            TomcatConnectionPool pool = poolMap.get(properties);
+            if (pool == null) {
+                synchronized (ClientPoolDriver.class) {
```
Take this sync above then you can remove a redundant pool == null check. Double check is enough.

kneeraj commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/ClientPoolDriver.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    private Properties getPoolProperties(String url) {
+        Properties properties = new Properties();
```
Just use empty. Let the pool use defaults. Check this. Let the pool handle the default case. Driver will be free from this.

kneeraj commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/TomcatConnectionPool.java:
```diff
@@ -0,0 +1,160 @@
+
+        List<String> listPoolPropKeys = Arrays.asList(poolPropKeys);
+
+        PoolProperties poolProperties = initiatePoolProperties(prop);
```
Change the name to initializePoolProperties

kneeraj reviewed (approved):
Approving for now. Suggested some cleanup. Please take care before merge to master.

Add unit tests too for stock spark 2.1.1 and at least spark 2.3

jramnara commented:
I don't understand why we changed the URL to use the pool. This should simply be through a JDBC connection property. The JDBC standard wants URL to conform to jdbc:[subprotocol]:[node]/[databaseName] ... so, pool is considered a sub-protocol ?

PradeepSurale committed:
incorporated the review comment.

Separated the connect() call for pool and the client.
Removed associated condition checks.

Commit:
Update bucket stats for SerializedDiskBuffer in sync block

Ensure that reference decrement and bucket stats update for a SerializedDiskBuffer
are atomic so that any concurrent reader possibly changing internal buffer from
compressed to decompressed and vice-versa is atomic for correct stats

PradeepSurale committed:
Updated Pool properties with default values

Also, removed the try catch, and deligating the exception to
the driver.

PradeepSurale commented:
Hi @jramnara the earlier implementation was the property based only, later on we internally had discussion and decided to go with this (url based) implementation. However, we could go to the properties based implementation.. @kneeraj @sumwale @dshirish

Commit:
Fix to SNAP-2578 and SNAP-2579

PradeepSurale committed:
Handled the case sensitivity in the URL.

sumwale commented:
@jramnara It is common to use the sub-protocol to denote major functionality when the underlying implementation itself behaves quite differently. For example in MySQL jdbc:mysql:replication:// or jdbc:mysql:loadbalance:// are used for relatively minor features like load balancing or server replication.  Or in Oracle jdbc:oracle:thin denotes the thin driver. There is no requirement for the sub-protocol to be database name as per standard (please point to standard if I am wrong) though some implementations have that interpretation.

Going by above my thinking was that having pool as part of sub-protocol will be cleaner. Especially since it refers to an entirely new implementation of the Driver so it is also cleaner to select the required implementation with the "Driver.acceptsURL" API. Nevertheless I don't have any strong opinion either way so if you think a property is better then lets have that.

jramnara commented:
ok. we can leave it the way it is. Looks like JDBC does permit sub-sub-protocol specification. But, to be "jdbc complaint" we should be using the jdbc DataSource conn pool API at some point.

sumwale commented:
@jramnara There are DataSource pool implementations in GemXD (see [ClientPooledConnection](https://github.com/SnappyDataInc/snappy-store/blob/147df2b05dfdeb06940583dd0df352da15f25971/gemfirexd/client/src/main/java/com/pivotal/gemfirexd/internal/client/ClientPooledConnection.java) and [ClientXAConnection](https://github.com/SnappyDataInc/snappy-store/blob/147df2b05dfdeb06940583dd0df352da15f25971/gemfirexd/client/src/main/java/io/snappydata/thrift/internal/datasource/ClientXAConnection.java)) that can be easily ported for thrift connections. However, that API does not do pooling itself rather defines a cleaner way for JavaEE to use pooling mechanisms on top of it like cleanup when connection is returned to pool. So the pooling work here is orthogonal to that, though later it could itself configure the tomcat/hikari pool implementations to use DataSource based impl rather than direct physical connections (in my testing earlier these did not make any difference to overall behaviour or performance).

dshirish commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/ClientPoolDriver.java:
```diff
@@ -0,0 +1,151 @@
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+/*
```
Since this is a new file do we need Pivotal notice?

dshirish commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/TomcatConnectionPool.java:
```diff
@@ -0,0 +1,168 @@
+        Set<String> keys = prop.stringPropertyNames();
+        String connectionProperties = keys.stream().filter(x -> listPoolPropKeys.contains(x))
+                .map(i -> i.toString() + "=" + prop.getProperty(i.toString()))
+                .collect(Collectors.joining(";"));
```
Since pool properties always start with "pool-", another way could to filter out properties stating with "pool-"

PradeepSurale replied:
Yes, however tested this with pool or random properties it wont be affecting anything.

dshirish commented on gemfirexd/client/src/main/resources/META-INF/services/java.sql.Driver:
```diff
@@ -1 +1 @@
-io.snappydata.jdbc.ClientDriver
+io.snappydata.jdbc.ClientDriver
```
Remove this file from the commit since there no change in it

PradeepSurale committed:
Handled the review comments

PradeepSurale committed:
Merge branch 'snappy/master' of https://github.com/SnappyDataInc/snappy-store into jdbc-pool-driver

PradeepSurale committed:
Merge branch 'jdbc-pool-driver' of https://github.com/SnappyDataInc/snappy-store into jdbc-pool-driver

sumwale commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/ClientPoolDriver.java:
```diff
@@ -0,0 +1,135 @@
+            return conn;
+        }
+    }
+}
```
Indentation of this class is incorrect. We use 2-spaces -- please setup codeStyleSettings.xml in IDEA as given in the docs.

PradeepSurale replied:
1. Added the codeStyleSettings.xml
2. Ctr+Sht+F works instead Ctr+I for indentation and formatting.

sumwale commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/ClientPoolDriver.java:
```diff
@@ -0,0 +1,135 @@
+ */
+public class ClientPoolDriver extends ClientDriver {
+
+    private final static String SUBPROTOCOL = "(thrift:pool:|pool:)+";
```
Why "+"? Should not be allowed more than once. Don't need the "thrift:" prefixed one, just use "pool:".

PradeepSurale replied:
Corrected it to use single occurrence "(pool:){1}"

sumwale commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/ClientPoolDriver.java:
```diff
@@ -0,0 +1,135 @@
+            java.sql.DriverManager.registerDriver(registeredDriver__);
+            if (driver != null) {
+                java.sql.DriverManager.deregisterDriver(driver);
+            }
```
This should not deregister existing driver rather only register itself separately. The base ClientDriver does it that way because it is a superset of ClientDRDADriver and handles both the protocols. This class does not seem to be able to handle all three protocols. This is probably the reason for the issue where multiple listing in java.sql.Driver were causing trouble.

PradeepSurale replied:
Removed the deregisteration of the existing driver, though the multiple listing in java.sql.Driver is behaving same way.

sumwale commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/ClientPoolDriver.java:
```diff
@@ -0,0 +1,135 @@
+                protocol.equalsIgnoreCase(SNAPPY_PROTOCOL))
+                : "thrift:pool:".equalsIgnoreCase(subProtocolGroup)
+                || "pool:".equalsIgnoreCase(subProtocolGroup);
+    }
```
The ":thrift" and ":drda" strangeness in ClientDriver was all due to backward compatibility where it supports the old DRDA protocol based on system property but also can be invoked explicitly with jdbc:snappydata:drda. Same way with thrift. And to top it all, jdbc:snappydata would mean DRDA in old rowstore product (no longer supported) while it means thrift in SnappyData. As of now thrift is the only supported one with SnappyData. All this complication is not required here so will suggest getting rid of all the cruft from this class. This method can always return true in this class.

PradeepSurale replied:
As implementing the java.sql.Driver instead of the ClientDriver / ClientDRDADriver we do not require this method now.

sumwale commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/ClientPoolDriver.java:
```diff
@@ -0,0 +1,135 @@
+
+    private final static String SUBPROTOCOL = "(thrift:pool:|pool:)+";
+    private final static Pattern PROTOCOL_PATTERN = Pattern.compile(URL_PREFIX_REGEX +
+            SUBPROTOCOL + "//.*", Pattern.CASE_INSENSITIVE);
```
Not need of URL_PREFIX_REGEX. This driver need not support jdbc:gemfirexd. Use SNAPPY_PROTOCOL.

PradeepSurale replied:
Corrected it use only snappy protocol.

sumwale commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/TomcatConnectionPool.java:
```diff
@@ -0,0 +1,168 @@
+    private static DataSource datasource;
+
+    static enum PoolProps {
+
```
Fix indentation.

sumwale commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/ClientPoolDriver.java:
```diff
@@ -0,0 +1,135 @@
+        }
+
+        properties = (properties == null) ? new Properties() : properties;
+        String clientDriverURL = url.toLowerCase().replace("pool:", "");
```
This will fail in case this string appears in hostname/password etc (replace will change all occurrences). Replace SNAPPY_PROTOCOL + SUBPROTOCOL with SNAPPY_PROTOCOL.

Also cannot do toLowerCase here. The URL may have password or other property names/values that are case-sensitive. Use a regex for case-insensitive replace. That way you can also use "^" or replaceFirst to be very sure that only the start is replaced. Like: PROTOCOL_PREFIX.matcher(url).replaceFirst(SNAPPY_PROTOCOL) where PROTOCOL_PREFIX is case-insensitive pre-compiled Pattern of SNAPPY_PROTOCOL + SUBPROTOCOL.

PradeepSurale replied:
replaced with following code:
String clientDriverURL = url.replaceFirst(PROTOCOL_PATTERN.matcher(url)
        .replaceFirst(SUBPROTOCOL), "");

sumwale commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/ClientPoolDriver.java:
```diff
@@ -0,0 +1,135 @@
+        properties.setProperty(TomcatConnectionPool.PoolProps.DRIVER_NAME.key, ClientDriver.class.getName());
+
+        TomcatConnectionPool p = poolMap.get(properties);
+        if (p != null) {
```
Is a separate poolMap really required? I think snappy-core's ConnectionPool also does that but I don't think it is required. Tomcat pool should be a "keyed" pool that will have its own map to return a connection satisfying those properties.

PradeepSurale replied:
poolMap is required to have maintain connection against the Keyed properties and the value will be pooled connections. However, this can be moved out from ClienPoolDriver class and moved into the TomcatConnectionPool class itself. Did this refactoring.  Here property is acting as "keyed" only.

sumwale replied:
I understand that, but see few problems with this. Firstly just noticed that dataSource is a static inside TomcatConnectionPool. That is certainly a bug if there are going to be multiple TomcatConnectionPool instances, so that must be made a member (preferably a final).

Secondly this poolMap can keep on increasing in size and will never be cleared. For instance if many users are connecting then each will have its own pool given that user/password properties will be different. This can keep on growing without any chance of cleanup of the poolMap itself. What we want is perhaps an overall limit on total connections across all pools in the poolMap and a way to push out older connections across the pools once the limit is reached.

Lastly remove a pool from map when a pool has zero active connections for a while.

Due to above issues was thinking if it is possible to make-do with a single pool instead of a map but does not seem possible. One option is to use JdbcInterceptors of tomcat pool for cleanup (https://tomcat.apache.org/tomcat-8.5-doc/jdbc-pool.html#org.apache.tomcat.jdbc.pool.JdbcInterceptor). A global atomic count tracks the total count of connections across pools and if there are too many then explicitly run testAllIdle() etc to force cleanup of idle connections, and then explicitly start closing non-active connections equally from all pools if it still is not enough.

Another option is to forgo tomcat-pool and instead use GenericKeyedObjectPool from commons-pool2 then don't need to worry about it and a single pool will do since its already keyed but then have to do own connection management.

sumwale replied:
Btw, we already use last one for P2P connections -- see ConnectionTable.connectionPool. That one already allows providing a connection key (which can be pool properties like you have for poolMap) and will track overall size/eviction etc. Thinking about it, the commons-pool2 might be enough for our uses since we don't need most of the extra features like statement caching/proxying etc. It looks like the existing ConnectionPool could also switch to it for the same reasons which can be tracked as a separate task.

What do you think about the available options? The basic issue being cleanup and preventing unbounded growth of poolMap (e.g. if poolMap grows to 100 and each pool to 100 then we reach 10000 connections or even more and no cleanup of pools themselves).

PradeepSurale replied:
Agree with the points you have mentioned the cleanup activity and continuous increasing of pool size is not that viable. Surely, I will go through the code points you have suggested.. (ConnectionTable.connectionPool etc)

PradeepSurale replied:
**1. For Cleanup:**
The tomcat jdbc pool itself having support for the cleaning up the Idle connection. If connection stays idle for specified duration then the properties minEvictableIdleTimeMillis (default;60secs) and timeBetweenEvictionRunsMillis are checks and the idle connection are removed from the pool.
For this I have tested it across the different pool( with different keys) for every pool, internally a separate eviction thread executed and cleaned up the idle connection. 
It removes all the connections which are more the value set with **minIdle**. If we set minIdle=1 it will keep only 1 idle connection in the pool.
So, what I think is we will not require a different monitoring thread, we simply have to tune this properties according to need.

**2. For Reset (autoCommit, readOnly and IsolationLevel)**
As suggested we have added the JDBC Interceptor for this. (also the scala test). Whenever the new connection is borrowed from the pool it resets the value to default and returns the connection.

sumwale commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/ClientPoolDriver.java:
```diff
@@ -0,0 +1,135 @@
+            }
+            Connection conn = pool.getConnection();
+            return conn;
+        }
```
Don't use double checked locking (you will need to make poolMap as volatile or final to work that correctly). With concurrent maps use putIfAbsent pattern or the new "compute" method with JDK8. Declare the type of poolMap to be ConcurrentHashMap or ConcurrentMap and:

Either
```java
else {
  TomcatConnectionPool pool = new TomcatConnectionPool(properties);
  TomcatConnectionPool oldPool = poolMap.putIfAbsent(properties, pool);
  if (oldPool != null) pool = oldPool;
  ...
}
```
Or better with JDK8 replace entire if with compute (notice the use of method reference which is cleaner for this case or else use lambda like: props -> new TomcatConnectionPool(props))
```java
TomcatConnectionPool p = poolMap.computeIfAbsent(properties, TomcatConnectionPool::new);
```

PradeepSurale replied:
Thanks, Good to know this computIfAbsent(. , .) method. After replacing this with the double check locking, code became very shorter and looking well.

sumwale commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/TomcatConnectionPool.java:
```diff
@@ -0,0 +1,168 @@
+        INIT_SIZE("pool-initialSize", "10"),
+        MAX_ACTIVE("pool-maxActive", "100"),
+        MIN_IDLE("pool-minIdle", "10"),  // Default value is derived from initialSize:10
+        MAX_IDLE("pool-maxIdle", "100"),  // Default value is maxActive:100
```
For defaults of maxIdle/maxActive, see ExternalStoreUtils.defaultMaxExternalPoolSize. That was tested to be overall better for different kinds of machines. Use similar here for defaults.

PradeepSurale replied:
using the suggested default values for maxActive and maxIdle.
defaultSize = String.valueOf(Math.max(256, Runtime.getRuntime().availableProcessors() * 8));

sumwale commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/TomcatConnectionPool.java:
```diff
@@ -0,0 +1,168 @@
+        REMOVE_ABANDONED("pool-removeAbandoned", "false"),
+        REMOVE_ABANDONED_TIMEOUT("pool-removeAbandonedTimeout", "60"),
+        TIME_BETWEEN_EVICTION_RUNS_MILLIS("pool-timeBetweenEvictionRunsMillis", "60000"),
+        MIN_EVICTABLE_IDLE_TIME_MILLIS("pool-minEvictableIdleTimeMillis", "60000");
```
Will need to see how these properties and their defaults will work. Have you done some testing? Overall have found testOnBorrow=true, validationInterval=10000 to work better else we get lost connections etc. See the overall defaults in ExternalStoreUtils.getAllPoolProperties (hikariCP=false is the tomcat case).

You may also want to add enums for testOnBorrow and validationInterval

PradeepSurale replied:
added the enum field for testOnBorrow and validationInterval. With default values are testOnBorrow=true, validationInterval=10000

sumwale commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/ClientPoolDriver.java:
```diff
@@ -0,0 +1,135 @@
+        properties = (properties == null) ? new Properties() : properties;
+        String clientDriverURL = url.toLowerCase().replace("pool:", "");
+        properties.setProperty(TomcatConnectionPool.PoolProps.URL.key, clientDriverURL);
+        properties.setProperty(TomcatConnectionPool.PoolProps.DRIVER_NAME.key, ClientDriver.class.getName());
```
break the line, exceeds 100 chars

PradeepSurale replied:
Corrected the line indentation.

sumwale commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/TomcatConnectionPool.java:
```diff
@@ -0,0 +1,168 @@
+    static enum PoolProps {
+
+        DRIVER_NAME("pool-driverClassName", ClientDriver.class.getName()),
+        URL("pool-url", ""), // Compulsory field user must provide
```
I don't think user should be allowed to override these two properties. They should be fixed for use with thrift ClientDriver URL.

PradeepSurale replied:
yes, we are not allowing user to override, even if user passes it in props, we are explicitly setting it to use thrift ClientDriver (in the ClientPoolDriver.connect() before calling Tomcat Pool)

sumwale commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/TomcatConnectionPool.java:
```diff
@@ -0,0 +1,168 @@
+        DRIVER_NAME("pool-driverClassName", ClientDriver.class.getName()),
+        URL("pool-url", ""), // Compulsory field user must provide
+        USER("pool-user", "APP"),
+        PASSWORD("pool-password", "APP"),
```
Convention in Spark properties is to use dots for separation. I think all new modules should use that convention rather than old gem/gemxd convention of hyphens.

PradeepSurale replied:
Corrected with '.' convention

sumwale commented on gemfirexd/client/src/main/java/io/snappydata/jdbc/TomcatConnectionPool.java:
```diff
@@ -0,0 +1,168 @@
+
+        String minEvictableIdleTimeMillis = prop.getProperty(PoolProps.MIN_EVICTABLE_IDLE_TIME_MILLIS.key,
+                PoolProps.MIN_EVICTABLE_IDLE_TIME_MILLIS.defValue);
+        poolProperties.setMinEvictableIdleTimeMillis(Integer.parseInt(minEvictableIdleTimeMillis));
```
break the long lines above (can see 2-3 which exceed 100 chars)

PradeepSurale replied:
Corrected

sumwale reviewed (commented):
@PradeepSurale 
Wanted to review this before merge but was too engaged elsewhere. Had mentioned it earlier today in the call. Would have been better for this to go with streaming_sink merge. Anyway have added a bunch of comments, some minor some major. Please take care of those or add your comments.

Also as part of https://github.com/SnappyDataInc/snappydata/pull/1170 it will be better to move these classes to SnappyData layer itself in the new snappy-jdbc module.

sumwale commented:
@PradeepSurale 
I will suggest that ClientPoolDriver should not extend ClientDriver rather just implement java.sql.Driver. This will avoid mixing of the two code paths since the old one has lot of cruft. You only need static block to register driver like in ClientDRDADriver and the implementation of acceptsURIL and connect like in this case. Rest all methods are not required.

PradeepSurale commented:
Yes, now ClientPoolDriver is implementing java.sql.Driver instead of the ClientDriver

PradeepSurale commented:
@sumwale @kneeraj @dshirish 
- What is the default values of following properties in our case?
1. **defaultAutoCommit** - true or false **?**
2. **defaultReadOnly**     -  true or false **?**
3. **defaultTransactionIsolation**  - TRANSACTION_NONE or 
                                                   TRANSACTION_READ_COMMITED or 
                                                    TRANSACTION_READ_UNCOMMITED or 
                                                    TRANSACTION_REPEATABLE_READ or
                                                     TRANSACTION_SERIALIZABLE
